blockbuster/proposals.go

```diff
@@ -0,0 +1,200 @@
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/skip-mev/pob/blockbuster/utils"
+)
+
+var _ BlockProposal = (*Proposal)(nil)
+
+type (
+	// LaneProposal defines the interface/APIs that are required for the proposal to interact
+	// with a lane.
+	LaneProposal interface {
+		// Logger returns the lane's logger.
+		Logger() log.Logger
+
+		// GetMaxBlockSpace returns the maximum block space for the lane as a relative percentage.
+		GetMaxBlockSpace() sdk.Dec
+
+		// Name returns the name of the lane.
+		Name() string
+	}
+
+	// BlockProposal is the interface/APIs that are required for proposal creation + interacting with
+	// and updating proposals. BlockProposals are iteratively updated as each lane prepares its
+	// partial proposal. Each lane must call UpdateProposal with its partial proposal in PrepareLane. BlockProposals
+	// can also include vote extensions, which are included at the top of the proposal.
+	BlockProposal interface {
+		// UpdateProposal updates the proposal with the given transactions. There are a
+		// few invarients that are checked:
+		//  1. The total size of the proposal must be less than the maximum number of bytes allowed.
+		//  2. The total size of the partial proposal must be less than the maximum number of bytes allowed for
+		//     the lane.
+		UpdateProposal(lane LaneProposal, partialProposalTxs [][]byte) error
+
+		// GetMaxTxBytes returns the maximum number of bytes that can be included in the proposal.
+		GetMaxTxBytes() int64
+
+		// GetTotalTxBytes returns the total number of bytes currently included in the proposal.
+		GetTotalTxBytes() int64
+
+		// GetTxs returns the transactions in the proposal.
+		GetTxs() [][]byte
+
+		// GetNumTxs returns the number of transactions in the proposal.
+		GetNumTxs() int
+
+		// Contains returns true if the proposal contains the given transaction.
+		Contains(tx []byte) bool
+
+		// AddVoteExtension adds a vote extension to the proposal.
+		AddVoteExtension(voteExtension []byte)
+
+		// GetVoteExtensions returns the vote extensions in the proposal.
+		GetVoteExtensions() [][]byte
+
+		// GetProposal returns all of the transactions in the proposal along with the vote extensions
+		// at the top of the proposal.
+		GetProposal() [][]byte
+	}
+
+	// Proposal defines a block proposal type.
+	Proposal struct {
+		logger log.Logger
+
+		// txs is the list of transactions in the proposal.
+		txs [][]byte
+
+		// voteExtensions is the list of vote extensions in the proposal.
+		voteExtensions [][]byte
+
+		// cache is a cache of the selected transactions in the proposal.
+		cache map[string]struct{}
+
+		// totalTxBytes is the total number of bytes currently included in the proposal.
+		totalTxBytes int64
+
+		// maxTxBytes is the maximum number of bytes that can be included in the proposal.
+		maxTxBytes int64
+	}
+)
+
+// NewProposal returns a new empty proposal.
+func NewProposal(maxTxBytes int64, logger log.Logger) *Proposal {
```

Comment:
Seems weird to me for such a type to have access to a logger.

Reply:
Agreed, I utilize the lane's logger instead. Removed this in the latest commit.